oak_functions/loader/src/server.rs

```diff
@@ -88,70 +86,91 @@ impl Policy {
-    /// the response equal to `self.constant_response_size_bytes`
-    ///
-    /// The zeros are added as the header value for header `oak-policy-padding`.
-    /// Returns an error if the size of the input `response` is larger than or equal to
-    /// `self.constant_response_size_bytes`.
-    fn pad_response(&self, response: FunctionsResponse) -> anyhow::Result<Response<Body>> {
-        let response_size = estimate_size(&response)?;
-        let required_padding = self.constant_response_size_bytes as i64 - response_size;
-        if required_padding < 0 {
-            panic!(
-                "Something is very wrong. Only call `pad_response` with responses that require padding."
-            )
-        } else {
-            let padding_bytes = vec![b'0'; required_padding as usize];
-            http::response::Builder::new()
-                .status(response.status_code)
-                .header(POLICY_PADDING_HEADER, std::str::from_utf8(&padding_bytes)?)
-                .body(response.body.map(Body::from).unwrap_or_else(Body::empty))
-                .context("couldn't create response")
-        }
-    }
 }
 
-/// The response from invoking the Wasm module functionality.
-pub struct FunctionsResponse {
-    pub status_code: StatusCode,
-    pub body: Option<Vec<u8>>,
+/// Creates an HTTP response, with status `OK`, and the protobuf-encoded `response` as its body.
+///
+/// To conform to the given security policy, the generated response is padded by adding a number of
+/// unspecified characters, to make the total size of the response equal to
+/// `policy.constant_response_size_bytes`. The padding is added as the value of the
+/// `oak-policy-padding` header. If the size of the input `response`, encoded as a protobuf message,
+/// is larger than `policy.constant_response_size_bytes` and the response is not already indicating
+/// policy violation, then the `response` is discarded and an HTTP response indicating the policy
+/// violation is returned. If `policy.constant_response_size_bytes` is too small even for the
+/// policy-violation response, or the response already indicates a policy violation, this function
+/// is recursively called to generate and return a response, where the protobuf-encoded body
+/// encapsulates `StatusCode::PolicyViolationInError` and an empty inner body. The recursion is
+/// guaranteed to terminate for a valid policy. If the policy is not valid, the function returns
+/// early with an error.
+fn create_response_and_apply_policy(
+    response: Response,
+    policy: &Policy,
+) -> anyhow::Result<http::response::Response<Body>> {
+    // check that the policy is valid
+    policy.validate()?;
+
+    // Serialize the response and get its size.
+    let mut encoded = vec![];
+    response
+        .encode(&mut encoded)
+        .context("couldn't encode response")?;
+    let padding_length =
+        i64::from(policy.constant_response_size_bytes) - i64::try_from(encoded.len()).unwrap();
```

Comment:
I think it gets encoded by hyper with HTTP1 because it only supports text, while HTTP2 supports binary. In any case it should not matter, the point is that it has a constant size, not that the size is exactly that specified in the policy. As long as it does not change based on the content, it is fine if the final size is a deterministic function of the "fixed size" value. Though this should probably be clarified.

Reply:
Just to clarify: are you saying that the existing implementation will provide a fixed size with HTTP1?

I don't believe it will. If the body is base64 encoded and the header is not, adding 1 byte to the header will not correctly offset  1 byte less in the body. Even if both are base64 encoded I don't think it will always be exact due to the 4 byte chunks that base64 encoding grow in.

Reply:
@conradgrobler you are right, I stand corrected. The encoding only applies to the body, and base64 encoding is not a linear operation, e.g. individually base64 encoding two separate bytes and concatenating the output is not the same as base64 encoding the two concatenated bytes. Thanks for pointing out.

Reply:
I think the simplest solution is to apply the size limit policy only on the size of the body bytes. It is easy to implement and easy to explain in documentation.

Now that we are always sending an OK status code to the client, the headers should not depend on the Wasm code at all. So this should result in a constant response size for all HTTP1.1 responses and a different constant response size for all the HTTP2 responses.

Reply:
To be clear, I mean the limit should be applied on the raw body bytes, not the base64-encoded bytes in the final HTTP 1.1 response.

Reply:
Thanks. I'll fix this in a separate PR.